src/Std/Data/DHashMap/Internal/RawLemmas.lean

```diff
@@ -2537,6 +2539,159 @@ end Const
+
+theorem alter_equiv_congr [LawfulBEq α] (h₁ : m₁.1.WF) (h₂ : m₂.1.WF) (h : m₁.1 ≈ m₂.1)
+    {k : α} (f : Option (β k) → Option (β k)) : (m₁.alter k f).1 ≈ (m₂.alter k f).1 := by
+  simp_to_model [Equiv, alter] using List.alterKey_of_perm _ h.1
+
+theorem modify_equiv_congr [LawfulBEq α] (h₁ : m₁.1.WF) (h₂ : m₂.1.WF) (h : m₁.1 ≈ m₂.1)
+    {k : α} (f : β k → β k) : (m₁.modify k f).1 ≈ (m₂.modify k f).1 := by
+  simp_to_model [Equiv, modify] using List.modifyKey_of_perm _ h.1
+
+theorem equiv_of_forall_get?_eq [LawfulBEq α] (h₁ : m₁.1.WF) (h₂ : m₂.1.WF) :
+    (∀ k, m₁.get? k = m₂.get? k) → m₁.1 ≈ m₂.1 := by
+  simp_to_model [get?, Equiv] using List.getValueCast?_ext
+
+namespace Const
+
+variable {β : Type v} (m₁ m₂ : Raw₀ α fun _ => β)
+variable [EquivBEq α] [LawfulHashable α]
+
+theorem get?_eq_of_equiv (h₁ : m₁.1.WF) (h₂ : m₂.1.WF) (h : m₁.1 ≈ m₂.1) {k : α} :
+    get? m₁ k = get? m₂ k := by
+  simp_to_model [Const.get?] using List.getValue?_of_perm _ h.1
+
+theorem get_eq_of_equiv (h₁ : m₁.1.WF) (h₂ : m₂.1.WF) (h : m₁.1 ≈ m₂.1)
+    {k : α} (h' : m₁.contains k) :
+    get m₁ k h' = get m₂ k ((contains_eq_of_equiv _ _ h₁ h₂ h).symm.trans h') := by
+  simp_to_model [Const.get] using List.getValue_of_perm _ h.1
+
+theorem get!_eq_of_equiv (h₁ : m₁.1.WF) (h₂ : m₂.1.WF) (h : m₁.1 ≈ m₂.1)
+    {k : α} [Inhabited β] : get! m₁ k = get! m₂ k := by
+  simp_to_model [Const.get!] using List.getValue!_of_perm _ h.1
+
+theorem getD_eq_of_equiv (h₁ : m₁.1.WF) (h₂ : m₂.1.WF) (h : m₁.1 ≈ m₂.1)
+    {k : α} {fallback : β} : getD m₁ k fallback = getD m₂ k fallback := by
+  simp_to_model [Const.getD] using List.getValueD_of_perm _ h.1
+
+omit [EquivBEq α] [LawfulHashable α] in
+
+theorem toList_perm_toList_of_equiv (h : m₁.1 ≈ m₂.1) :
+    (Raw.Const.toList m₁.1).Perm (Raw.Const.toList m₂.1) := by
+  simp_to_model [Const.toList]
+  exact h.1.map _
+
+theorem insertMany_list_equiv_congr (h₁ : m₁.1.WF) (h₂ : m₂.1.WF) (h : m₁.1 ≈ m₂.1)
+    {l : List (α × β)} : (insertMany m₁ l).1.1 ≈ (insertMany m₂ l).1.1 := by
+  simp_to_model [Equiv, Const.insertMany] using List.insertList_perm_of_perm_first h.1
+
+theorem insertManyIfNewUnit_list_equiv_congr {m₁ m₂ : Raw₀ α fun _ => Unit}
+    (h₁ : m₁.1.WF) (h₂ : m₂.1.WF) (h : m₁.1 ≈ m₂.1) {l : List α} :
+    (insertManyIfNewUnit m₁ l).1.1 ≈ (insertManyIfNewUnit m₂ l).1.1 := by
+  simp_to_model [Equiv, Const.insertManyIfNewUnit]
+    using List.insertListIfNewUnit_perm_of_perm_first h.1
+
+theorem alter_equiv_congr (h₁ : m₁.1.WF) (h₂ : m₂.1.WF) (h : m₁.1 ≈ m₂.1)
+    {k : α} (f : Option β → Option β) : (alter m₁ k f).1 ≈ (alter m₂ k f).1 := by
+  simp_to_model [Equiv, Const.alter] using List.Const.alterKey_of_perm _ h.1
+
+theorem modify_equiv_congr (h₁ : m₁.1.WF) (h₂ : m₂.1.WF) (h : m₁.1 ≈ m₂.1)
+    {k : α} (f : β → β) : (modify m₁ k f).1 ≈ (modify m₂ k f).1 := by
+  simp_to_model [Equiv, Const.modify] using List.Const.modifyKey_of_perm _ h.1
```

Comment:
How about adding lemmas for `filterMap`, `filter` and `map`? I think the relevant theory should be there.